### PR TITLE
Add PropertyForSale and PropertyForRent. 

### PR DESCRIPTION
**Pull request by jinnsuke**
Add supporting functionality such as storage, parsing etc

### CODE REVIEW

File: src/main/java/seedu/address/model/AddressBook.java [hunk 1]
```diff
@@ -8,30 +8,31 @@
...
+ * Wraps all data at the address-book level.
+ * Duplicates are not allowed (by .isSamePerson comparison).
  */
 public class AddressBook implements ReadOnlyAddressBook {
 
     private final UniquePersonList persons;
+    private final UniquePropertyForSaleList propertiesForSale; // New list for properties for sale
+    private final UniquePropertyForRentList propertiesForRent; // New list for properties for rent
 
-    /*
```
Comment (starbucksventi) [inline src/main/java/seedu/address/model/AddressBook.java:20]: I think this comment (lines 20-26) is needed, it was by the original creators of the address book

File: src/main/java/seedu/address/model/AddressBook.java [hunk 2]
```diff
@@ -68,44 +85,66 @@ public boolean hasPerson(Person person) {
...
+        requireNonNull(newData);
+
+        setPersons(newData.getPersonList());
+        setPropertiesForSale(newData.getPropertyForSaleList());
+        setPropertiesForRent(newData.getPropertyForRentList());
     }
 
+    //// person-level operations
+
+    // Existing methods for managing persons...
```
Comment (starbucksventi) [inline src/main/java/seedu/address/model/AddressBook.java]: Might want to delete this comment (116)

File: src/main/java/seedu/address/model/property/UniquePropertyForRentList.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+     * The property must exist in the list.
+     */
+    public void remove(PropertyForRent toRemove) {
+        requireNonNull(toRemove);
+        if (!internalList.remove(toRemove)) {
+            throw new PropertyNotFoundException();
+        }
+    }
+
+    public void setProperties(UniquePropertyForRentList replacement) {
```
Comment (starbucksventi) [inline src/main/java/seedu/address/model/property/UniquePropertyForRentList.java:82]: Might need JavaDoc for this

File: src/main/java/seedu/address/model/property/UniquePropertyForSaleList.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+            return false;
+        }
+
+        UniquePropertyForSaleList otherUniquePropertyList = (UniquePropertyForSaleList) other;
+        return internalList.equals(otherUniquePropertyList.internalList);
+    }
+
+    @Override
+    public int hashCode() {
+        return internalList.hashCode();
```
Comment (starbucksventi) [inline src/main/java/seedu/address/model/property/UniquePropertyForSaleList.java:124]: Might want to +internalUnmodifiableList.hashCode() too

File: src/main/java/seedu/address/model/property/UniquePropertyForRentList.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            return false;
+        }
+
+        UniquePropertyForRentList otherUniquePropertyList = (UniquePropertyForRentList) other;
+        return internalList.equals(otherUniquePropertyList.internalList);
+    }
+
+    @Override
+    public int hashCode() {
+        return internalList.hashCode();
```
Comment (starbucksventi) [inline src/main/java/seedu/address/model/property/UniquePropertyForRentList.java:129]: Might want to add internalUnmodifiableList.hashCode(), similar to what you did for previous parts

File: src/main/java/seedu/address/model/property/exceptions/DuplicatePropertyException.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package seedu.address.model.property.exceptions;
+
+/**
+ * Signals that the operation will result in duplicate properties (properties are
+ * considered duplicates if they have the same identity).
+ */
+public class DuplicatePropertyException extends RuntimeException {
+    public DuplicatePropertyException() {
+        super("Operation would result in duplicate properties");
```
Comment (starbucksventi) [inline src/main/java/seedu/address/model/property/exceptions/DuplicatePropertyException.java:9]: Assign the string to a constant and use it (avoid magic strings)

File: src/main/java/seedu/address/storage/JsonAdaptedPropertyForRent.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+        if (town == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, "town"));
+        }
+
+        if (propertyType == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, "propertyType"));
+        }
+
+        if (size <= 0) {
```
Comment (starbucksventi) [inline src/main/java/seedu/address/storage/JsonAdaptedPropertyForRent.java:82]: Need to transfer the validity checks to static methods in respective class (e.g. Phone.isValidPhone(phone) in JsonAdaptedPerson.java)
Need to do this for all 4 conditionals here

File: src/main/java/seedu/address/storage/JsonAdaptedPropertyForRent.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    /**
+     * Converts this Jackson-friendly adapted property object into the model's {@code PropertyForRent} object.
+     *
+     * @throws IllegalValueException if there were any data constraints violated in the adapted property.
+     */
+    public PropertyForRent toModelType() throws IllegalValueException {
+        if (address == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, "address"));
```
Comment (starbucksventi) [inline src/main/java/seedu/address/storage/JsonAdaptedPropertyForRent.java:71]: The second argument in String.format needs to follow the Name.class.getSimple__() format, similar to JsonAdaptedPerson.java (for all 3 conditionals)

File: src/main/java/seedu/address/storage/JsonAdaptedPropertyForRent.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        if (town == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, "town"));
+        }
+
+        if (propertyType == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, "propertyType"));
+        }
+
+        if (size <= 0) {
+            throw new IllegalValueException("Size must be greater than zero.");
```
Comment (starbucksventi) [inline src/main/java/seedu/address/storage/JsonAdaptedPropertyForRent.java:83]: All the the strings here need to be transferred to constants. See JsonAdaptedPerson.java for reference

File: src/main/java/seedu/address/storage/JsonAdaptedPropertyForSale.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    /**
+     * Converts this Jackson-friendly adapted property object into the model's {@code PropertyForSale} object.
+     *
+     * @throws IllegalValueException if there were any data constraints violated in the adapted property.
+     */
+    public PropertyForSale toModelType() throws IllegalValueException {
+        if (address == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, "address"));
```
Comment (starbucksventi) [inline src/main/java/seedu/address/storage/JsonAdaptedPropertyForSale.java:70]: Same comment as above (second argument in String.format needs to follow the Name.class.getSimple__() format, similar to JsonAdaptedPerson.java (for all 3 conditionals)) for "address", "town", "propertyType" and "listingDate"

File: src/main/java/seedu/address/storage/JsonAddressBookStorage.java [hunk 11]
```diff
@@ -4,77 +4,214 @@
...
 import seedu.address.commons.exceptions.DataLoadingException;
 import seedu.address.commons.exceptions.IllegalValueException;
 import seedu.address.commons.util.FileUtil;
 import seedu.address.commons.util.JsonUtil;
 import seedu.address.model.ReadOnlyAddressBook;
+import seedu.address.model.property.PropertyForRent;
+import seedu.address.model.property.PropertyForSale;
 
 /**
- * A class to access AddressBook data stored as a json file on the hard disk.
```
Comment (starbucksventi) [inline src/main/java/seedu/address/storage/JsonAddressBookStorage.java:18]: Change back to the previous JavaDoc comment if this was accidental

File: src/main/java/seedu/address/storage/JsonAddressBookStorage.java [hunk 12]
```diff
@@ -4,77 +4,214 @@
...
+
+    /**
+     * Reads properties for rent from the JSON file and returns a list of PropertyForRent.
+     *
+     * @return a list of PropertyForRent read from the file.
+     * @throws DataLoadingException if there is an error loading properties for rent.
+     */
+    public List<PropertyForRent> readPropertiesForRent() throws DataLoadingException {
+        JsonSerializableAddressBook jsonAddressBook = JsonUtil.readJsonFile(filePath, JsonSerializableAddressBook.class)
+                .orElseThrow(() -> new RuntimeException("Failed to load properties for rent"));
```
Comment (starbucksventi) [inline src/main/java/seedu/address/storage/JsonAddressBookStorage.java:162]: Same as previous, turn strings into constants for this and the 2 other strings in the method

File: src/main/java/seedu/address/storage/JsonAddressBookStorage.java [hunk 13]
```diff
@@ -4,77 +4,214 @@
...
+
+    /**
+     * Reads properties for sale from the JSON file and returns a list of PropertyForSale.
+     *
+     * @return a list of PropertyForSale read from the file.
+     * @throws DataLoadingException if there is an error loading properties for sale.
+     */
+    public List<PropertyForSale> readPropertiesForSale() throws DataLoadingException {
+        JsonSerializableAddressBook jsonAddressBook = JsonUtil.readJsonFile(filePath, JsonSerializableAddressBook.class)
+                .orElseThrow(() -> new RuntimeException("Failed to load properties for sale"));
```
Comment (starbucksventi) [inline src/main/java/seedu/address/storage/JsonAddressBookStorage.java:185]: Put strings into constants (all 3)

File: src/main/java/seedu/address/storage/JsonSerializableAddressBook.java [hunk 14]
```diff
@@ -12,6 +12,8 @@
 import seedu.address.model.AddressBook;
 import seedu.address.model.ReadOnlyAddressBook;
 import seedu.address.model.person.Person;
+import seedu.address.model.property.PropertyForRent; // Import your PropertyForRent model
```
Comment (starbucksventi) [inline src/main/java/seedu/address/storage/JsonSerializableAddressBook.java]: Remove the comments  on 15-16

File: src/test/java/seedu/address/logic/commands/AddCommandTest.java [hunk 15]
```diff
@@ -84,90 +81,6 @@ public void toStringMethod() {
         assertEquals(expected, addCommand.toString());
     }
 
-    /**
```
Comment (starbucksventi) [inline src/test/java/seedu/address/logic/commands/AddCommandTest.java:87]: Wait why was this deleted (87-170)

**Review by starbucksventi — CHANGES_REQUESTED** [review]

File: src/test/java/seedu/address/logic/commands/AddCommandTest.java (hunk 15 above)
Comment (jinnsuke) [reply src/test/java/seedu/address/logic/commands/AddCommandTest.java:87]: I created a new class file for this because I need to use the Model stub for my property stuff as well
